File: frontend/src/api/knowledge.ts
import client from './client'
import type { KnowledgeItem, PaginatedResponse, Tag } from '@/types'

export interface FetchKnowledgeParams {
  page?: number
  per_page?: number
  search?: string
  source_type?: string
  status?: string
  tags?: string
}

export async function fetchKnowledge(
  params: FetchKnowledgeParams = {}
): Promise<PaginatedResponse<KnowledgeItem>> {
  const p: Record<string, string | number> = {}
  if (params.page !== undefined) p.page = params.page
  if (params.per_page !== undefined) p.per_page = params.per_page
  if (params.search?.trim()) p.search = params.search.trim()
  if (params.source_type && params.source_type !== 'all') p.source_type = params.source_type
  if (params.status && params.status !== 'all') p.status = params.status
  if (params.tags?.trim()) p.tags = params.tags.trim()

  const res = await client.get<PaginatedResponse<KnowledgeItem>>('/api/knowledge', { params: p })
  return res.data
}

export async function fetchKnowledgeItem(id: string): Promise<KnowledgeItem> {
  const res = await client.get<KnowledgeItem>(`/api/knowledge/${id}`)
  return res.data
}

export async function deleteKnowledgeItem(id: string): Promise<void> {
  await client.delete(`/api/knowledge/${id}`)
}

export async function patchKnowledgeItem(id: string, updates: { tags?: string[]; title?: string }): Promise<KnowledgeItem> {
  const res = await client.patch<KnowledgeItem>(`/api/knowledge/${id}`, updates)
  return res.data
}

export async function fetchTags(): Promise<Tag[]> {
  const res = await client.get<Tag[]>('/api/tags')
  return res.data
}

File: backend/api/knowledge.py
"""
GET    /api/knowledge        — paginated, filterable list
GET    /api/knowledge/{id}   — full item
PATCH  /api/knowledge/{id}   — update tags / notes
DELETE /api/knowledge/{id}   — delete item
"""

import json
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from backend.db.engine import get_db_session
from backend.db.repositories.knowledge import KnowledgeItemRepository

router = APIRouter(prefix="/knowledge", tags=["knowledge"])


class KnowledgeItemUpdate(BaseModel):
    tags: Optional[list[str]] = None
    title: Optional[str] = None


@router.get("")
def list_knowledge(
    source_type: Optional[str] = None,
    status: Optional[str] = None,
    search: Optional[str] = None,
    tags: Optional[str] = None,   # comma-separated
    page: int = 1,
    per_page: int = 20,
    db: Session = Depends(get_db_session),
):
    tag_list = [t.strip() for t in tags.split(",")] if tags else None
    items, total = KnowledgeItemRepository.list_all(
        db,
        source_type=source_type,
        status=status,
        search=search,
        tags=tag_list,
        page=page,
        page_size=per_page,
    )
    return {
        "items": [_serialize(item) for item in items],
        "total": total,
        "page": page,
        "per_page": per_page,
    }


@router.get("/{item_id}")
def get_knowledge_item(item_id: str, db: Session = Depends(get_db_session)):
    item = KnowledgeItemRepository.get_by_id(db, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    return _serialize(item, include_content=True)


@router.patch("/{item_id}")
def update_knowledge_item(
    item_id: str,
    body: KnowledgeItemUpdate,
    db: Session = Depends(get_db_session),
):
    updates = {}
    if body.tags is not None:
        updates["tags"] = json.dumps(body.tags)
    if body.title is not None:
        updates["title"] = body.title
    if not updates:
        raise HTTPException(status_code=400, detail="No fields to update")
    item = KnowledgeItemRepository.update(db, item_id, updates)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    return _serialize(item)


@router.delete("/{item_id}")
def delete_knowledge_item(item_id: str, db: Session = Depends(get_db_session)):
    ok = KnowledgeItemRepository.delete(db, item_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Item not found")
    return {"ok": True}


def _serialize(item, include_content: bool = False) -> dict:
    meta = item.youtube_metadata
    d = {
        "id": item.id,
        "source_type": item.source_type,
        "source_id": item.source_id,
        "title": item.title,
        "author": item.author,
        "published_at": item.published_at.isoformat() if item.published_at else None,
        "ingested_at": item.ingested_at.isoformat() if item.ingested_at else None,
        "summary": item.summary,
        "summary_length": item.summary_length,
        "tags": _parse_json(item.tags, []),
        "topics": _parse_json(item.topics, {}),
        "llm_model": item.llm_model,
        "word_count": item.word_count,
        "status": item.status,
        "error_message": item.error_message,
        # YouTube-specific metadata (nullable for non-youtube sources)
        "channel": meta.channel if meta else None,
        "views": meta.views if meta else None,
        "duration": meta.duration if meta else None,
        "subscribers": meta.subscribers if meta else None,
        "videos_count": meta.videos_count if meta else None,
        "thumbnail_url": meta.thumbnail_url if meta else None,
        "detected_language": meta.detected_language if meta else None,
        "timestamps": _parse_json(meta.timestamps, {}) if meta else {},
    }
    if include_content:
        d["raw_content"] = item.raw_content
    return d


def _parse_json(value, default):
    if not value:
        return default
    try:
        return json.loads(value)
    except Exception:
        return default
